Instrumenting your code with Go AST - Blog post (#407)
mattermost/mattermost-developer-documentation: site/content/blog/2019-10-25-instrumenting-go-code-via-ast.md

---

> title: "Instrumenting Go via AST"

Reviewer: Not sure if you meant the title to be this, given the slug below?

```suggestion
title: "Instrumenting Go code via AST"
```

slug: instrumenting-go-code-via-ast
date: 2019-10-25T00:00:00-04:00
author: Eli Yukelzon
github: reflog
community: eli.yukelzon
---

We've been working on integrating call tracing in the server to provide exact measurements of all API and DB calls.

> We've picked [OpenTracing](https://github.com/opentracing/opentracing-go) - a lovely open source project that allows you to setup trace reporting and enables you to support [Distributed tracing](https://opentracing.io/docs/overview/what-is-tracing/)

Reviewer 2: ```suggestion
We've picked [OpenTracing](https://github.com/opentracing/opentracing-go) - a lovely open source project that allows you to setup trace reporting and enables you to support [Distributed tracing](https://opentracing.io/docs/overview/what-is-tracing/).
```

> Instrumenting your API handler in Go is very straightforward - setup connection to a collection server supporting OpenTracing (we've decided to use [Jaeger](https://www.jaegertracing.io/)) spec and wrap you code in spans.

Reviewer: ```suggestion
Instrumenting your API handler in Go is very straightforward - setup a connection to a collection server supporting the OpenTracing spec (we've decided to use [Jaeger](https://www.jaegertracing.io/)) and wrap your code in spans.
``

> Instrumenting your API handler in Go is very straightforward - setup connection to a collection server supporting OpenTracing (we've decided to use [Jaeger](https://www.jaegertracing.io/)) spec and wrap you code in spans.

Reviewer 2: ```suggestion
Instrumenting your API handler in Go is very straightforward - just set up a connection to a collection server supporting OpenTracing (we've decided to use [Jaeger](https://www.jaegertracing.io/)) spec and wrap your code in spans.
```

To simplify this, we've added a simple tracing module:
```go
package tracing

import (
	opentracing "github.com/opentracing/opentracing-go"
	"github.com/uber/jaeger-lib/metrics"

	"github.com/uber/jaeger-client-go"
	jaegercfg "github.com/uber/jaeger-client-go/config"
	jaegerlog "github.com/uber/jaeger-client-go/log"

> 	"context"

Reviewer: I feel a little obsessive saying this, but it would be amazing if these were alphabetically ordered and grouped like we do with other imports :)

                Param: 1,
        },
        Reporter: &jaegercfg.ReporterConfig{
                LogSpans: true,
        },
    }

    _, err := cfg.InitGlobalTracer(
        "mattermost",
        jaegercfg.Logger(jaegerlog.StdLogger),
        jaegercfg.Metrics(metrics.NullFactory),
    )
    if err != nil {
        return err
    }

    initialized = true

    return nil
}

func StartRootSpanByContext(ctx context.Context, operationName string) (opentracing.Span, context.Context) {
    return opentracing.StartSpanFromContext(ctx, operationName)
}

func StartSpanWithParentByContext(ctx context.Context, operationName string) (opentracing.Span, context.Context) {
    parentSpan := opentracing.SpanFromContext(ctx)

    if parentSpan == nil {
        return StartRootSpanByContext(ctx, operationName)
    }

    return opentracing.StartSpanFromContext(ctx, operationName, opentracing.ChildOf(parentSpan.Context()))
}
```

Now, wrapping an API call in a span is very easy:

```go

func apiCall(c *Context, w http.ResponseWriter, r *http.Request) {
	span, ctx := tracing.StartSpanWithParentByContext(c.App.Context, "api4:apiCall")
	c.App.Context = ctx

>     defer span.Finish()

Reviewer 3: nit: spurious extra space before `defer` here
```suggestion
   defer span.Finish()
```

    // perform actual request handling
}
```

Now, each time the `apiCall` handler is invoked, it will be measured and traced on Jaeger.

> The problem is - we have a rather large API surface (around 300+ handlers) and it would be a really tough task to instrument all the handlers by hand. That's where our story begins.

Reviewer: ```suggestion
The problem is: we have a rather large API surface (around 300+ handlers) and it would be a really tough task to instrument all the handlers by hand. That's where our story begins.
```

> The problem is - we have a rather large API surface (around 300+ handlers) and it would be a really tough task to instrument all the handlers by hand. That's where our story begins.

Reviewer 2: ```suggestion
The problem is that we have a rather large API surface (around 300+ handlers) and it would be a really tough task to instrument all the handlers by hand. That's where our story begins.
```

## Go AST

> So what is AST really? Well, to quote Wikipedia:

Reviewer: ```suggestion
So what is an AST really? Well, to quote Wikipedia:
```

> So what is AST really? Well, to quote Wikipedia:

Reviewer 2: You could add a link to the Wikipedia page.

> In computer science, an abstract syntax tree (AST), or just syntax tree, is a tree representation of the abstract syntactic structure of source code written in a programming language. Each node of the tree denotes a construct occurring in the source code.

![ast example](https://upload.wikimedia.org/wikipedia/commons/thumb/c/c7/Abstract_syntax_tree_for_Euclidean_algorithm.svg/1024px-Abstract_syntax_tree_for_Euclidean_algorithm.svg.png?1572188936106)

> Basically, AST is a tree like representation of your source code.

Reviewer: ```suggestion
Basically, an AST is a tree-like representation of your source code.
```

> Basically, AST is a tree like representation of your source code.

Reviewer 2: ```suggestion
Basically, AST is a tree-like representation of your source code.
```

> ![ast example](https://upload.wikimedia.org/wikipedia/commons/thumb/c/c7/Abstract_syntax_tree_for_Euclidean_algorithm.svg/1024px-Abstract_syntax_tree_for_Euclidean_algorithm.svg.png?1572188936106)

Reviewer 3: Should we host this file ourselves instead, in case it breaks in the future?

Author: good idea

Why do we need it?

> Parsing AST allows us to refactor code on a statement level, meaning, instead of finding methods and fields using string search, we can find them by their actual structure. This gives us incredible flexibility in writing custom go refactoring tools.

Reviewer: ```suggestion
Parsing the code into an AST allows us to refactor code on a statement level, meaning, instead of finding methods and fields using string search, we can find them by their actual structure. This gives us incredible flexibility in writing custom go refactoring tools.
```

Let's see a small example for a 'Hello World' program:

```go
package main

import (
	"fmt"
)

func main() {
	fmt.Printf("Hello, Golang\n")
}

```

> After passing it through `go/parser` we get the following structure (image generated using: http://goast.yuroyoro.net/)

Reviewer 2: ```suggestion
After passing it through `go/parser`, we get the following structure (image generated using: http://goast.yuroyoro.net/):
```

![hello ast](/blog/2019-10-25-instrumenting-go-code-via-ast/hello_ast.png)

> So given this structure, we can easily find for example the `Printf` statement by looking at `File.Decls[1].Body.List[0]`

Reviewer 2: ```suggestion
So given this structure, we can easily find, for example, the `Printf` statement by looking at `File.Decls[1].Body.List[0]`.
```

> Now that we understand the basic idea behind AST, let's go ahead and try to solve the issue at hand - instrumenting our handlers with tracing code.

Reviewer: ```suggestion
Now that we understand the basic idea behind ASTs, let's go ahead and try to solve the issue at hand - instrumenting our handlers with tracing code.
```

## Analyzing existing code

### Finding the pattern

> First thing we need is to identify the methods that we want to instrument. In our code base, all of the API handlers sit in `api4` folder and each method that handles a certain API has the same signature:

Reviewer: ```suggestion
The first thing we need is to identify the methods that we want to instrument. In our code base, all of the API handlers sit in the `api4` folder and each method that handles a certain API has the same signature:
```

```go
func apiCall(c *Context, w http.ResponseWriter, r *http.Request) {
    // perform actual request handling
}
```

So we are looking at a function that has **exactly** 3 arguments, with specific names and types.

> Let's write AST walker boilerplate that will go through all the files in `api4` folder and parse them:

Reviewer: ```suggestion
Let's write some AST walker boilerplate that will go through all the files in the `api4` folder and parse them:
```

```go
package main

import (
	"bytes"
	"fmt"
	"go/ast"
	"go/format"
	"go/parser"
	"go/token"
	"io/ioutil"
)

func fix(dir string) {
	fset := token.NewFileSet()
	pkgs, err := parser.ParseDir(fset, dir, nil, parser.ParseComments)
	if err != nil {
		panic(err)
	}

	for _, pkg := range pkgs {
		for fileName, file := range pkg.Files {
			fmt.Printf("working on file %v\n", fileName)
			ast.Inspect(file, func(n ast.Node) bool {
                // perform analysis here
				return true
			})

			buf := new(bytes.Buffer)
			err := format.Node(buf, fset, file)
			if err != nil {
				fmt.Printf("error: %v\n", err)
			} else if fileName[len(fileName)-8:] != "_test.go" {
				ioutil.WriteFile(fileName, buf.Bytes(), 0664)
			}
		}
	}
}

func main() {
	fix("./api4")
}
```

> `ast.Inspect` will visit all AST nodes in each file and provides us with a parsed tokens.

Reviewer 2: ```suggestion
`ast.Inspect` visits all AST nodes in each file and provides us with a parsed token.
```

> `ast.Inspect` will visit all AST nodes in each file and provides us with a parsed tokens.

Reviewer: ```suggestion
`ast.Inspect` will visit all AST nodes in each file and provides us with the parsed tokens.
```

> To only operate on specific nodes in the AST, we try to cast to the type we are interested in and only the proceed:

Reviewer 2: ```suggestion
To only operate on specific nodes in the AST, we try to cast to the type we are interested in and only then proceed:
```

> To only operate on specific nodes in the AST, we try to cast to the type we are interested in and only the proceed:

Reviewer: ```suggestion
To only operate on specific nodes in the AST, we try to cast to the type we are interested in and only then proceed:
```

```go
fn, ok  := n.(*ast.FuncDecl)
if ok {
  // current node is a function!
}
```

> Next we need to extract and analyze the parameters to the function (we'll also add a small helper function to get node as string):

Reviewer: ```suggestion
Next we need to extract and analyze the parameters to the function (we'll also add a small helper function to get a node as a string):
```

```go
func FormatNode(node ast.Node) string {
	buf := new(bytes.Buffer)
	_ = format.Node(buf, token.NewFileSet(), node)
	return buf.String()
}
// retreive function's parameter list
params  := fn.Type.Params.List
// we are only interested in functions with exactly 3 parameters
if  len(params) ==  3 {

> 	first_parameter_is_c := FormatNode(params[0].Names[0]) ==  "c"  &&  FormatNode(params[0].Type) ==  "*Context"

Reviewer 3: Perhaps unrelated to the change, but why do we enforce the parameter name here -- presumably, if an API happened to use `ctx` instead, we might miss it, no?

Author: Correct. But I wanted to be super strict in my search pattern...

	second_parameter_is_w := FormatNode(params[1].Names[0]) ==  "w"  &&  FormatNode(params[1].Type) ==  "http.ResponseWriter"  
	third_parameters_is_r := FormatNode(params[2].Names[0]) ==  "r"  &&  FormatNode(params[2].Type) ==  "*http.Request"
	if first_parameter_is_c && second_parameter_is_w && third_parameters_is_r {
		// this is an API handler!
	}
}
```

> Now that we've found our function, the fun begins! We want to add a couple of statements as described in the introduction and a relevant import

Reviewer 2: ```suggestion
Now that we've found our function, the fun begins! We want to add a couple of statements as described in the introduction and a relevant import.
```

## Instrumenting the code

> As a reminder, this is the piece of code we want to inject in the begining of every API handler to instrument the code:

Reviewer 2: ```suggestion
As a reminder, this is the piece of code we want to inject at the beginning of every API handler to instrument the code:
```

```go
	span, ctx := tracing.StartSpanWithParentByContext(c.App.Context, "api4:apiCall")
	c.App.Context = ctx
    defer span.Finish()
```

First of all - we need to take care of imports. We are using the 'tracing' module so we need to add just one line (utilizing the golang.org/x/tools/go/ast/astutil module):
```go
astutil.AddImport(fset, file, "github.com/mattermost/mattermost-server/services/tracing")
```

> Now we are ready to inject the code. To do that, we need to convert it from it's textual representation into a set of AST nodes.

Reviewer: ```suggestion
Now we are ready to inject the code. To do that, we need to convert it from its textual representation into a set of AST nodes.
```

// c.App.Context = ctx							
a2 := ast.AssignStmt{
    // token.ASSIGN is =
	Tok: token.ASSIGN,
	Lhs: []ast.Expr{
		&ast.SelectorExpr{
			X: &ast.SelectorExpr{
				X:   &ast.Ident{Name: "c"},
				Sel: &ast.Ident{Name: "App"},
			},
			Sel: &ast.Ident{Name: "Context"},
		},
	},
	Rhs: []ast.Expr{
		&ast.Ident{Name: "ctx"},
	},
}
// last statement is 'defer'							
a3 := ast.DeferStmt{
	// what function call should be deferred?
	Call: &ast.CallExpr{
	    // Finish from 'span' identifier
		Fun: &ast.SelectorExpr{
			X:   &ast.Ident{Name: "span"},
			Sel: &ast.Ident{Name: "Finish"},
		},
	},
}
// now we prepend the three statements before the rest of function body							
fn.Body.List = append([]ast.Stmt{&a1, &a2, &a3}, fn.Body.List...)
```

And that's it! We have our custom refactoring/instrumentation tool that we can tweak however we want.

## Go2AST

Since the process of writing AST code by hand based on `go/printer` or [http://goast.yuroyoro.net/](http://goast.yuroyoro.net/) is rather tedious, I've taken the code of `go/printer` and converted it into a reusable CLI command that does all the work for you! Just feed it some Go code and it'll spit out a ready-to-paste AST tree. 

> Take a look [https://github.com/reflog/go2ast](https://github.com/reflog/go2ast)

Reviewer 2: ```suggestion
Take a look here: [https://github.com/reflog/go2ast](https://github.com/reflog/go2ast).
```

## Conclusion

> Writing refactoring tools is fun and possibilities are endless. Go standard library contains everything you need to create your own little tools for every repeatable task.

Reviewer 2: ```suggestion
Writing refactoring tools is fun and the possibilities are endless. Go standard library contains everything you need to create your own little tools for every repeatable task.
```

> Writing refactoring tools is fun and possibilities are endless. Go standard library contains everything you need to create your own little tools for every repeatable task.

Reviewer: ```suggestion
Writing refactoring tools is fun and the possibilities are endless. The Go standard library contains everything you need to create your own little tools for every repeatable task.
```

> Seeing code's AST at the first time can be a little daunting, but once you get this knowledge in your tool-belt, you'll look for other places to apply it immediately! 

Reviewer 2: ```suggestion
Seeing code's AST for the first time can be a little daunting, but once you get this knowledge in your tool-belt, you'll look for other places to apply it immediately! 
```

> Seeing code's AST at the first time can be a little daunting, but once you get this knowledge in your tool-belt, you'll look for other places to apply it immediately! 

Reviewer: ```suggestion
Seeing code as an AST for the first time can be a little daunting, but once you get this knowledge in your tool-belt, you'll look for other places to apply it immediately! 
```

---
title: "Instrumenting Go code via AST"
slug: instrumenting-go-code-via-ast
date: 2019-10-25T00:00:00-04:00
author: Eli Yukelzon
github: reflog
community: eli.yukelzon
---

We've been working on integrating call tracing in the server to provide exact measurements of all API and DB calls.
We've picked [OpenTracing](https://github.com/opentracing/opentracing-go) - a lovely open source project that allows you to setup trace reporting and enables you to support [Distributed tracing](https://opentracing.io/docs/overview/what-is-tracing/).

Instrumenting your API handler in Go is very straightforward - setup a connection to a collection server supporting the OpenTracing spec (we've decided to use [Jaeger](https://www.jaegertracing.io/)) and wrap your code in spans.

> ``

Reviewer 2: ```suggestion

```

---
title: "Instrumenting Go code via AST"
slug: instrumenting-go-code-via-ast
date: 2019-10-25T00:00:00-04:00
author: Eli Yukelzon
github: reflog
community: eli.yukelzon
---

We've been working on integrating call tracing in the server to provide exact measurements of all API and DB calls.
We've picked [OpenTracing](https://github.com/opentracing/opentracing-go) - a lovely open source project that allows you to setup trace reporting and enables you to support [Distributed tracing](https://opentracing.io/docs/overview/what-is-tracing/).

Instrumenting your API handler in Go is very straightforward - setup a connection to a collection server supporting the OpenTracing spec (we've decided to use [Jaeger](https://www.jaegertracing.io/)) and wrap your code in spans.
``

To simplify this, we've added a simple tracing module:
```go
package tracing

import (
	opentracing "github.com/opentracing/opentracing-go"
	"github.com/uber/jaeger-lib/metrics"

    "context"

	"github.com/uber/jaeger-client-go"
	jaegercfg "github.com/uber/jaeger-client-go/config"
	jaegerlog "github.com/uber/jaeger-client-go/log"	

> )

Reviewer: Not sure the inline suggestion is working properly with the multi-line comment, but basically it was still bugging me so I ran it through `gofmt` to get them in the right order

```suggestion
import (
	"context"

	opentracing "github.com/opentracing/opentracing-go"
	"github.com/uber/jaeger-client-go"
	jaegercfg "github.com/uber/jaeger-client-go/config"
	jaegerlog "github.com/uber/jaeger-client-go/log"
	"github.com/uber/jaeger-lib/metrics"
)
```

---
title: "Instrumenting Go code via AST"
slug: instrumenting-go-code-via-ast
date: 2019-10-25T00:00:00-04:00
author: Eli Yukelzon
github: reflog
community: eli.yukelzon
---

We've been working on integrating call tracing in the server to provide exact measurements of all API and DB calls.
We've picked [OpenTracing](https://github.com/opentracing/opentracing-go) - a lovely open source project that allows you to setup trace reporting and enables you to support [Distributed tracing](https://opentracing.io/docs/overview/what-is-tracing/).

Instrumenting your API handler in Go is very straightforward - setup a connection to a collection server supporting the OpenTracing spec (we've decided to use [Jaeger](https://www.jaegertracing.io/)) and wrap your code in spans.
``

To simplify this, we've added a simple tracing module:
```go
package tracing

import (
	opentracing "github.com/opentracing/opentracing-go"
	"github.com/uber/jaeger-lib/metrics"

    "context"

	"github.com/uber/jaeger-client-go"
	jaegercfg "github.com/uber/jaeger-client-go/config"
	jaegerlog "github.com/uber/jaeger-client-go/log"	

> import (

Reviewer: Yeah the multiline suggest didn't work :'(
